### ISSUE TITLE
Add upwind flux dispatch with stencil-width-aware mask fallback for non-rectangular domains

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/finitevolX#68]
## Description

`finitevolX` has `ArakawaCGridMask` with `StencilCapability` that computes how many stencil points are available at each grid point, and has `Reconstruction1D/2D/3D` with various WENO schemes. However, there is **no runtime flux function that automatically dispatches** the correct stencil width near irregular domain boundaries — falling back from high-order (e.g. 6-point WENO) to lower-order (4-point, then 2-point linear) based on local mask information.

Both reference repos implement this critical dispatch pattern:

### Reference: `louity/qgsw-pytorch` (mask-weighted blending)

**[`src/flux.py`](https://github.com/louity/qgsw-pytorch/blob/673c8e60288baaf1e426dfc3ddd4266380f781cf/src/flux.py)** — The `flux()` function computes all stencil widths simultaneously and blends them using mutually exclusive mask arrays (`mask_2`, `mask_4`, `mask_6`):

```python
def flux(q, u, dim, n_points,
         rec_func_2, rec_func_4, rec_func_6,
         mask_2, mask_4, mask_6):
    u_pos = F.relu(u)
    u_neg = u - u_pos

    # 2-point reconstruction (everywhere)
    q_stencil2 = stencil_2pts(q, dim)
    qi2_pos = rec_func_2(*q_stencil2)
    qi2_neg = rec_func_2(*q_stencil2[::-1])

    # 4-point reconstruction (padded to full size)
    q_stencil4 = stencil_4pts(q, dim)
    qi4_pos = F.pad(rec_func_4(*q_stencil4), pad)
    qi4_neg = F.pad(rec_func_4(*q_stencil4[::-1]), pad)

    # 6-point reconstruction (padded to full size)
    q_stencil6 = stencil_6pts(q, dim)
    qi6_pos = F.pad(rec_func_6(*q_stencil6), pad)
    qi6_neg = F.pad(rec_func_6(*q_stencil6[::-1]), pad)

    # Blend using masks
    return u_pos * (mask_2*qi2_pos + mask_4*qi4_pos + mask_6*qi6_pos) \
         + u_neg * (mask_2*qi2_neg + mask_4*qi4_neg + mask_6*qi6_neg)
```

**[`src/sw.py`](https://github.com/louity/qgsw-pytorch/blob/673c8e60288baaf1e426dfc3ddd4266380f781cf/src/sw.py#L201-L250)** — Shows how the masks are connected:
```python
self.h_flux_y = lambda h, v: flux(h, v, dim=-1, n_points=6,
    rec_func_2=linear2_centered,
    rec_func_4=wenoz4_left,
    rec_func_6=wenoz6_left,
    mask_2=self.masks.v_sten_hy_eq2[...,1:-1],
    mask_4=self.masks.v_sten_hy_eq4[...,1:-1],
    mask_6=self.masks.v_sten_hy_gt6[...,1:-1])
```

### Reference: `louity/MQGeometry` (distance-to-boundary masks)

**[`masks.py`](https://github.com/louity/MQGeometry/blob/1511b1a9bbc4b1213017f59305e2ce466b308169/masks.py#L19-L82)** — Computes `distbound1`, `distbound2`, `distbound3plus` for each variable (u, v, psi, q) using average-pooling to detect proximity to boundaries.

**[`flux.py`](https://github.com/louity/MQGeometry/blob/1511b1a9bbc4b1213017f59305e2ce466b308169/flux.py)** — `div_flux_5pts_mask` uses `distbound1/2/3+` to cat together 1-pt, 3-pt, and 5-pt stencil reconstructions.

## Proposed API

```python
def upwind_flux(
    q: Array,                        # tracer field (cell-centered)
    u: Array,                        # transport velocity (face-centered)
    dim: int,                        # dimension for flux computation
    max_stencil: int = 6,            # max stencil width (2, 4, or 6)
    rec_funcs: dict = None,          # {2: linear2, 4: weno4, 6: weno6}
    mask_hierarchy: dict = None,     # {2: mask_2pt, 4: mask_4pt, 6: mask_6pt}
) -> Array:
    """Compute upwind flux with automatic stencil fallback near boundaries.

    Uses mask_hierarchy (from ArakawaCGridMask.stencil_capability) to blend
    reconstructions of different widths. Near boundaries where only 2 points
    are available, falls back to linear; in the interior, uses full WENO.
    """
```

Should integrate with existing `ArakawaCGridMask.stencil_capability` to generate the mask hierarchy.

## Acceptance Criteria

- [ ] `upwind_flux` function with mask-based stencil dispatch
- [ ] Support for 2-point, 4-point, and 6-point stencils
- [ ] Integration with `ArakawaCGridMask` to auto-generate mask hierarchy
- [ ] Tests on rectangular domain (no fallback needed — should match pure WENO)
- [ ] Tests on irregular/masked domain verifying correct fallback near boundaries
- [ ] Tests verifying conservation properties (total tracer preserved)
- [ ] JAX-compatible (no Python control flow on data; use `jnp.where` or mask blending)

## Priority
**Medium-High** — critical for non-rectangular ocean geometries. Your `ArakawaCGridMask` already has the mask data, but no flux function uses it.

---

_References:_
- _[louity/qgsw-pytorch/src/flux.py](https://github.com/louity/qgsw-pytorch/blob/673c8e60288baaf1e426dfc3ddd4266380f781cf/src/flux.py)_
- _[louity/qgsw-pytorch/src/sw.py](https://github.com/louity/qgsw-pytorch/blob/673c8e60288baaf1e426dfc3ddd4266380f781cf/src/sw.py#L201-L250)_
- _[louity/MQGeometry/masks.py](https://github.com/louity/MQGeometry/blob/1511b1a9bbc4b1213017f59305e2ce466b308169/masks.py)_
- _[louity/MQGeometry/flux.py](https://github.com/louity/MQGeometry/blob/1511b1a9bbc4b1213017f59305e2ce466b308169/flux.py)_
